### PR TITLE
support or, and, implies, distinct in mbp_basic. fix #6865 #6866

### LINKED ISSUE #6865: ASSERTION VIOLATION, File: ../src/qe/mbp/mbp_basic_tg.cpp, Line: 60
Commit: https://github.com/Z3Prover/z3/commit/a8c4384536019e438adb7a8df6d24e93d3431ab6
OS: Ubuntu 22.04

```
[552] % z3debug small.smt2
ASSERTION VIOLATION
File: ../src/qe/mbp/mbp_basic_tg.cpp
Line: 60
!m.is_or(term)
(C)ontinue, (A)bort, (S)top, (T)hrow exception, Invoke (G)DB
a
[553] % cat small.smt2
(declare-const a Bool)
(declare-const b Bool)
(assert (exists ((c Bool) (d Bool) (e Bool)) (and e (or (= a c) (= b d)) (= a c) (= b d))))
(apply qe2)
```

### PR DESCRIPTION
**Pull request by hgvk94**

### CODE REVIEW

File: src/qe/mbp/mbp_basic_tg.cpp [hunk 1]
```diff
@@ -43,32 +43,106 @@ struct mbp_basic_tg::impl {
...
+            if (is_seen(term)) continue;
+            if (m.is_ite(term, c, th, el) && should_split(c)) {
                 mark_seen(term);
                 progress = true;
                 if (m_mdl.is_true(c)) {
                     m_tg.add_lit(c);
                     m_tg.add_eq(term, th);
-                }
-                else {
+                } else {
```
Comment (NikolajBjorner) [inline src/qe/mbp/mbp_basic_tg.cpp:78]: Elsewhere I try to standardize code to have } and else on separate lines.

File: src/qe/mbp/mbp_basic_tg.cpp [hunk 2]
```diff
@@ -43,32 +43,106 @@ struct mbp_basic_tg::impl {
...
+                }
+            }
+            if (m_use_mdl && m.is_distinct(term)) {
+                mark_seen(term);
+                progress = true;
+                bool eq = false;
+                app *c = to_app(term);
+                for (auto a1 : *c) {
+                    for (auto a2 : *c) {
+                        if (a1 == a2) continue;
```
Comment (NikolajBjorner) [inline src/qe/mbp/mbp_basic_tg.cpp:132]: It is OK to have single line statement after if on a separate line. The code base isn't consistent about this but tends to mostly breaking a line after if.

File: src/qe/mbp/mbp_qel.cpp [hunk 3]
```diff
@@ -178,7 +178,7 @@ class mbp_qel::impl {
 
         std::function<bool(expr *)> non_core = [&](expr *e) {
             if (is_app(e) && is_partial_eq(to_app(e))) return true;
-            if (m.is_ite(e)) return true;
+            if (m.is_ite(e) || m.is_or(e) || m.is_implies(e) || m.is_and(e) || m.is_distinct(e)) return true;
```
Comment (NikolajBjorner) [inline src/qe/mbp/mbp_qel.cpp:181]: is_xor? is_iff? (is_iff is shorthand for is_eq and arguments are Boolean). is_distinct also takes non-boolean arguments so not sure what the discriminator is here.

File: src/qe/mbp/mbp_basic_tg.cpp (hunk 1 above)
Comment (hgvk94) [reply src/qe/mbp/mbp_basic_tg.cpp:78]: Oh. I ran clang format with the formatter for Spacer. I will update the formal file.

**Pull request merged**